plugins/json_ld.ts

```diff
@@ -0,0 +1,102 @@
+import { merge } from "../core/utils/object.ts";
+import { getDataValue } from "../core/utils/data_values.ts";
+
+import type Site from "../core/site.ts";
+import type { Page } from "../core/file.ts";
+import { Graph, Thing } from "npm:schema-dts@1.1.2";
+export interface Options {
+  /** The list extensions this plugin applies to */
+  extensions?: string[];
+
+  /** The key name for the transformations definitions */
+  name?: string;
+}
+
+const defaults: Options = {
+  extensions: [".html"],
+  name: "json_ld",
+};
+export type JsonldData = Graph | Thing;
+
+// Helper function to determine if a value is an object
+function isObject(value: unknown) {
+  return value && typeof value === "object" && !Array.isArray(value);
+}
+
+/**
+ * A plugin to insert structured JSON-LD data for SEO and social media
+ * @see https://lume.land/plugins/json_ld/
+ */
+export function json_ld(userOptions?: Options) {
+  const options = merge(defaults, userOptions);
+
+  return (site: Site) => {
+    site.mergeKey(options.name, "object");
+    site.process(options.extensions, (pages) => pages.forEach(json_ld));
+
+    function json_ld(page: Page) {
+      let json_ld = page.data[options.name] as JsonldData | undefined;
+
+      if (!json_ld || !page.document) {
+        return;
+      }
+      const { document, data } = page;
+
+      // Recursive function to traverse and process JSON-LD data
+      function traverse(key: string | undefined, value: unknown): unknown {
+        if (typeof value === "string") {
+          const data_value = getDataValue(data, value);
+          // Check if the value is a URL or ID that needs to be processed
+          if (
+            (key == "@id" || key == "url") &&
+            (value.startsWith("#") || value.startsWith("/"))
+          ) {
+            return site.url(data_value, true);
+          }
+          return data_value && data_value != "" ? data_value : undefined;
+        } else if (Array.isArray(value)) {
```

Comment:
The `else` is not neccesary because the prevous if always return something. (this is just a code style suggestion, I know :))